src/traces/histogram/calc.js

```diff
@@ -180,10 +157,181 @@ module.exports = function calc(gd, trace) {
+            };
+        }
+
+        // Even non-autobinned traces get included here, so we get the greatest extent
+        // and minimum bin size of them all.
+        // But manually binned traces won't be adjusted, even if the auto values
+        // are inconsistent with the manual ones (or the manual ones are inconsistent
+        // with each other).
+        //
+        // TODO: there's probably a weird case here where a larger bin pushes the
+        // start/end out, then it gets shrunk and doesn't make sense with the smaller bin.
+        // Need to look for cases like this and see if the results are acceptable
+        // or we need to think harder about it.
+        minSize = getMinSize(minSize, binspec.size);
+        minStart = Math.min(minStart, pa.r2c(binspec.start, 0, calendar));
+        maxEnd = Math.max(maxEnd, pa.r2c(binspec.end, 0, calendar));
+
+        // add the flag that lets us abort autobin on later traces
+        if(i) trace._autoBinFinished = 1;
+    }
+
+    // do what we can to match the auto bins to the first manual bins
+    // but only if sizes are all numeric
+    if(firstManual && isNumeric(firstManual.size) && isNumeric(minSize)) {
+        // first need to ensure the bin size is the same as or an integer fraction
+        // of the first manual bin
+        // allow the bin size to increase just under the autobin step size to match,
+        // (which is a factor of 2 or 2.5) otherwise shrink it
+        if(minSize > firstManual.size / 1.9) minSize = firstManual.size;
+        else minSize = firstManual.size / Math.ceil(firstManual.size / minSize);
+
+        // now decrease minStart if needed to make the bin centers line up
+        var adjustedFirstStart = firstManual.start + (firstManual.size - minSize) / 2;
+        minStart = adjustedFirstStart - minSize * Math.ceil((adjustedFirstStart - minStart) / minSize);
+    }
+
+    // now go back to the autobinned traces and update their bin specs with the final values
+    for(i = 0; i < autoBinnedTraces.length; i++) {
+        tracei = autoBinnedTraces[i];
+        calendar = tracei[maindata + 'calendar'];
+
+        tracei._input[binAttr] = tracei[binAttr] = {
+            start: pa.c2r(minStart, 0, calendar),
+            end: pa.c2r(maxEnd, 0, calendar),
+            size: minSize
+        };
+
+        // note that it's possible to get here with an explicit autobin: false
+        // if the bins were not specified.
+        // in that case this will remain in the trace, so that future updates
+        // which would change the autobinning will not do so.
+        tracei._input[autoBinAttr] = tracei[autoBinAttr];
+    }
+
+    return trace[binAttr];
+}
+
+/*
+ * return an array of traces that are all stacked or grouped together
+ * TODO: only considers histograms. Should we also harmonize with bars?
+ * in principle people can mix and match these, but bars always
+ * specify their positions explicitly...
```

Comment:
We *could* try to treat bars the same as we treat manually-binned histograms in adjusting the autobins. That has more edge cases to consider though, nonuniform bar spacing being perhaps the hardest to manage, so I'm inclined not to worry about this one either. If someone is grouping/stacking bars with histograms I'd like to think it would be clear to them that they need to set the histogram binning to match the bar positions...